site/app/libraries/homework/Gateways/Library/FileSystemLibraryGateway.php

```diff
@@ -0,0 +1,82 @@
+<?php namespace app\libraries\homework\Gateways\Library;
+
+
+use ZipArchive;
+use app\libraries\FileUtils;
+use app\libraries\homework\Gateways\LibraryGateway;
+
+class FileSystemLibraryGateway implements LibraryGateway {
+    const SUCCESS = 0;
+    const STDERR = 2;
+
+    protected function createFolderIfNotExists(string $path): bool {
+        return FileUtils::createDir($path);
+    }
+
+    /** @inheritDoc */
+    public function addGitLibrary(string $repoUrl, string $location): string {
+        if (!$this->createFolderIfNotExists($location)) {
+            return 'Error when creating folder.';
+        }
+
+        $sanitizedRepoUrl = escapeshellarg($repoUrl);
+        $sanitizedLocation = escapeshellarg($location);
+
+        $cmd = "git clone $sanitizedRepoUrl $sanitizedLocation";
+
+        $descriptors = [
+            ["pipe", "r"],  // stdin
+            ["pipe", "w"],  // stdout
+            ["pipe", "w"],  // stderr
+        ];
+        $git = proc_open($cmd, $descriptors, $pipes);
```

Comment:
Where exactly is `$pipes` coming from? Is this just some PHP global?

From the [Docs](https://www.php.net/manual/en/function.proc-open.php) it seems `$pipes` is just set automatically... Is this good practice?

Reply:
What? PHP globals are all caps and start with an underscore. `$pipes` is defined to be the array you see right before this. I'm just setting up descriptors for stdin, stdout, and stderr, but I only really care about stderr. I think it is just fine to be explicit about what I want. `$pipes` is an array telling PHP that I want to pipe it into memory streams that I can manipulate, and that the process should read from stdin, and write to stdout and stderr. We can also write to the programs stdin if we need to.

Reply:
Disregard my last comment on this. I didn't look to see exactly what you were referencing. The function proc_open fills out that variable for me. It's a pass by reference thing, and sicne it doesn't exist yet, it creates it and fills out the actual pipe descriptors for me based upon my requested descriptors that I passed in. So it opens them and then that's how it returns them, but it also needs to return the process handle, so it can't just return them. Hence the need for the pass by reference.

TL;DR:
I'm creating `$pipes` right here and proc_open is filling it out for me because it's a pass by reference.